### DISCUSSION TITLE
Annotation Hierarchy/Responses Available?

### DISCUSSION THREAD

**Discussion by rahul-evenup-ai** [pymupdf/PyMuPDF#1440] [Looking for help] [answered]
I see that you're able to clear all annotations that are in response to a primary annotation (delete_responses), but I was wondering if there was a way of grabbing all responses to an annotation, or really some way of linking a set of "response" annotations to a primary one. Thanks for the help!

**Comment by JorjMcKie** [comment]
No, sorry, this is currently not possible.

**Comment by JorjMcKie** [comment]
@rahul-evenup-ai - it might be an idea though to do this:

PopUps for a given annotation can be handled already.
One might parse the the page's annotation array and collect all `/IRT` annotations addressing the given one. Then put their xref in some list.
Using low-level code, this is actually already possible using Python code only ...

**Comment by JorjMcKie** [comment]
Made a little snippet to explain my somewhat cryptic previous post:
[irt-chain.zip](https://github.com/pymupdf/PyMuPDF/files/7693013/irt-chain.zip)

Produces this output:
```python irt-chain.py
XREF 17 (Highlight)
Popup XREF 18
Responses:
    XREF 13 (Text)
    XREF 15 (Text)
    XREF 19 (Text)
```

**Reply by JorjMcKie** [reply]
Because annotations can be loaded by xref as well, it should be possible to access every of the above (except the popup) directly.

**Comment by JorjMcKie** [comment]
Based on your suggestion, I am currently testing the following feature:
1. Annotation have an additional property ``Annot.irt_xref`` which (if > 0) points to an annotation to which this one is a response.
2. ``Annot.set_irt_xref(xref)`` sets this annotation as a response to another annotation identified by the argument.

**Comment by rahul-evenup-ai** [comment]
wow thanks so much @JorjMcKie, this is all really helpful. Will look into the zip file you shared with me, thanks so much for the help so far! Look forward to testing the `irt_xref` functionality when its ready 🙏

**Reply by JorjMcKie** [reply]
If you tell me your configuration, I can point you to a location of pre-release wheels: OS, Python X.Y, bitness (if  Windows).

**Reply by rahul-evenup-ai** [reply]
mac 12.0.1, Python 3.8.12

thanks!

**Reply by JorjMcKie** [reply]
[Here](https://github.com/pymupdf/PyMuPDF-wheels/tree/master/wheelhouse) are the current OSX wheels.

**Comment by JorjMcKie** [comment, marked as answer]
Implemented in v1.19.3.

**Reply by rahul-evenup-ai** [reply]
thank you so much, confirming that its working as intended for us! 🙏

**Comment by ghyman-oreilly** [comment]
Hi Jorj, when I try accessing Annot.irt_xref in v1.23.26 (and a few versions prior), I get a bound method error. Is this a bug or has the intended usage changed since 1.19.3? I am able to access the attribute when I use 1.19.3.

**Reply by JorjMcKie** [reply]
There was a typo unfortunately. I fixed it yesterday, so it will be included in the next version.

**Reply by ghyman-oreilly** [reply]
Thank you, Jorj! I appreciate what you do. It's a wonderful library.
